src/Microsoft.AspNetCore.Rewrite/Internal/ApacheModRewrite/ApacheModRewriteRule.cs

```diff
@@ -20,36 +22,33 @@ public ApacheModRewriteRule(UrlMatch initialMatch, IList<Condition> conditions,
             Actions = urlActions;
         }
 
-        public virtual async Task ApplyRuleAsync(RewriteContext context)
+        public virtual Task ApplyRuleAsync(RewriteContext context)
         {
             // 1. Figure out which section of the string to match for the initial rule.
             var initMatchRes = InitialMatch.Evaluate(context.HttpContext.Request.Path, context);
 
             if (!initMatchRes.Success)
             {
                 context.Logger?.ModRewriteDidNotMatchRule();
-                return;
+                return TaskCache.CompletedTask;
             }
 
-            BackReferenceCollection condBackReferences = null;
+            MatchResults condResult = null;
             if (Conditions != null)
             {
-                var condResult = ConditionHelper.Evaluate(Conditions, context, initMatchRes.BackReferences);
+                condResult = ConditionHelper.Evaluate(Conditions, context, initMatchRes.BackReferences);
                 if (!condResult.Success)
                 {
                     context.Logger?.ModRewriteDidNotMatchRule();
-                    return;
+                    return TaskCache.CompletedTask;
                 }
             }
 
             // At this point, we know our rule passed, first apply pre conditions,
             // which can modify things like the cookie or env, and then apply the action
             context.Logger?.ModRewriteMatchedRule();
 
-            foreach (var action in Actions)
-            {
-                await action.ApplyActionAsync(context, initMatchRes?.BackReferences, condBackReferences);
-            }
+            return Task.WhenAll(Actions.Select(action => action.ApplyActionAsync(context, initMatchRes?.BackReferences, condResult?.BackReferences)));
```

Comment:
Does this run the actions in parallel? That's not safe. This method probably still needs async/await.